client/client.go

```diff
@@ -1,29 +1,401 @@
 	"github.com/datastax/go-cassandra-native-protocol/frame"
+	"github.com/datastax/go-cassandra-native-protocol/message"
+	"github.com/datastax/go-cassandra-native-protocol/primitive"
+	"github.com/rs/zerolog/log"
+	"io"
+	"math"
 	"net"
+	"sync"
+	"sync/atomic"
+	"time"
 )
 
+const (
+	DefaultConnectTimeout = time.Second * 5
+	DefaultReadTimeout    = time.Second * 12
+)
+
+const (
+	DefaultMaxInFlight = 1024
+	DefaultMaxPending  = 10
+)
+
+const ManagedStreamId int16 = 0
+
+// CqlClient is a client for Cassandra-compatible backends. It is preferable to create CqlClient instances using the
+// constructor function NewCqlClient. Once the client is created and properly configured, use Connect or ConnectAndInit
+// to establish new connections to the server.
 type CqlClient struct {
-	address string
-	codec   frame.RawCodec
+	// The remote contact point address to connect to.
+	RemoteAddress string
+	// The AuthCredentials for authenticated servers. If nil, no authentication will be used.
+	Credentials *AuthCredentials
+	// The frame.Codec to use; if none provided, a default codec will be used.
+	Codec frame.Codec
+	// The maximum number of in-flight requests to apply for each connection created with Connect. Must be strictly
+	// positive.
+	MaxInFlight int
+	// The maximum number of pending responses awaiting delivery to store per request. Must be strictly positive.
+	// This is only useful when using continuous paging, a feature specific to DataStax Enterprise.
+	MaxPending int
+	// The timeout to apply when establishing new connections.
+	ConnectTimeout time.Duration
+	// The timeout to apply when waiting for incoming responses.
+	ReadTimeout time.Duration
 }
 
-func NewCqlClient(address string, codec frame.RawCodec) *CqlClient {
+// Creates a new CqlClient with default options. Leave credentials nil to opt out from authentication.
+func NewCqlClient(remoteAddress string, credentials *AuthCredentials) *CqlClient {
 	return &CqlClient{
-		address: address,
-		codec:   codec,
+		RemoteAddress:  remoteAddress,
+		Credentials:    credentials,
+		MaxInFlight:    DefaultMaxInFlight,
+		MaxPending:     DefaultMaxPending,
+		ConnectTimeout: DefaultConnectTimeout,
+		ReadTimeout:    DefaultReadTimeout,
 	}
 }
 
-func (client *CqlClient) Connect() (*CqlConnection, error) {
-	c, err := net.Dial("tcp", client.address)
-	if err != nil {
+func (client *CqlClient) String() string {
+	return fmt.Sprintf("CQL client [%v]", client.RemoteAddress)
+}
+
+// Connect establishes a new TCP connection to the client's remote address. The returned CqlClientConnection is ready
+// to use, but one must initialize it manually, for example by calling CqlClientConnection.InitiateHandshake.
+// Alternatively, use ConnectAndInit to get a fully-initialized connection.
+func (client *CqlClient) Connect() (*CqlClientConnection, error) {
+	log.Debug().Msgf("%v: connecting", client)
+	if conn, err := net.DialTimeout("tcp", client.RemoteAddress, client.ConnectTimeout); err != nil {
```

Comment:
We should use `DialContext` instead as contexts allow for more use cases besides timeouts (like cancellation).

Reply:
But we don't have a context at hand here? The dial operation already takes a timeout, I think that's enough.

Reply:
Yes, we should receive a context as parameter. A timeout is not enough for our use case, we require cancellation support.

Reply:
Actually, nevermind, we just need to expose the `newCqlClientConnection` method, because we create the `net.Conn` object ourselves.

Reply:
See https://github.com/datastax/go-cassandra-native-protocol/pull/9#discussion_r519745017